btree/btree.c

```diff
@@ -3166,21 +3178,199 @@ int64_t btree_destroy_tick(struct m0_sm_op *smop)
+			lev->l_seq = lev->l_node->n_seq;
+			oi->i_rc = node_find(&node_slot, &bop->bo_rec.r_key);
+			lev->l_idx = node_slot.s_idx;
+
+			if (node_level(node_slot.s_node) > 0) {
+				node_child(&node_slot, &child);
+				if (!address_is_valid(child)) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, M0_ERR(-EFAULT));
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child, P_NEXTDOWN);
+			} else
+				return P_LOCK;
+		} else {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
+		}
+	case P_LOCK:
+		if (!locked(tree))
+			return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
+					    bop->bo_arbor->t_desc, P_CHECK);
+		else
+			return P_CHECK;
+	case P_CHECK:
+		oi->i_trial++;
+		if (!path_check(oi, tree, &bop->bo_rec.r_key.k_cookie)) {
+			if (oi->i_trial == MAX_TRIALS) {
+				if (bop->bo_flags & OF_LOCKALL)
+					return fail(bop, -ETOOMANYREFS);
+				else
+					bop->bo_flags |= OF_LOCKALL;
+			}
+			if (bop->bo_arbor->t_height <= tree->t_height) {
+				/* If height increased */
+				//return m0_sm_op_sub(&bop->bo_op, P_CLEANUP,
+				//                    P_INIT);
+				lock_op_unlock(tree);
+				return P_CLEANUP;
+			} else {
+				/* If height decreased */
+				lock_op_unlock(tree);
+				return P_DOWN;
+			}
+		}
+		/** Fall through if path_check is successful. */
+	case P_ACT: {
+		m0_bcount_t  ksize;
+		m0_bcount_t  vsize;
+		void        *pkey;
+		void        *pval;
+		struct slot  s = {};
+		int	     rc;
+
+		lev = &oi->i_level[oi->i_used];
+
+		s.s_node	     = lev->l_node;
+		s.s_idx		     = lev->l_idx;
+		s.s_rec.r_key.k_data = M0_BUFVEC_INIT_BUF(&pkey, &ksize);
+		s.s_rec.r_val	     = M0_BUFVEC_INIT_BUF(&pval, &vsize);
+		s.s_rec.r_flags      = 0;
+		/**
+		 *  There are two cases based on the flag set by user :
+		 *  1. Flag BRF_EQUAL: If requested key found return record else
+		 *  return key not exist.
+		 *  2. Flag BRF_SLANT: If the key index(found during P_NEXTDOWN)
+		 *  is less than total number of keys, return the record at key
+		 *  index. Else loop through the levels to find valid sibling.
+		 *  If valid sibling found, return first key of the sibling
+		 *  subtree else return key not exist.
+		 */
+		if (bop->bo_rec.r_flags & BRF_EQUAL) {
+			if (oi->i_rc)
+				node_rec(&s);
+			else
+				s.s_rec.r_flags |= BRF_NOENT;
+		} else {
+			if (lev->l_idx < node_count(lev->l_node))
+				node_rec(&s);
+			else {
+				rc = btree_sibling_first_key_get(oi, tree, &s);
+				if (rc != 0) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, rc);
+				}
+			}
+		}
+
+		bop->bo_cb.c_act(&bop->bo_cb, &s.s_rec);
+
+		lock_op_unlock(tree);
+		return P_CLEANUP;
+		//return m0_sm_op_sub(&bop->bo_op, P_CLEANUP, P_DONE);
 	}
+	case P_CLEANUP: {
+		int i;
+		for (i = 0; i < oi->i_used; ++i) {
+			if (oi->i_level[i].l_node != NULL) {
+				node_put(oi->i_level[i].l_node);
+				oi->i_level[i].l_node = NULL;
+			}
+			if (oi->i_level[i].l_alloc != NULL) {
```

Comment:
I do not see l_alloc allocated so this freeing up is not necessary.

Reply:
Ok, will remove l_alloc.